### PR TITLE
ENH add imagenet

### PR DESCRIPTION
**Pull request by zaccharieramzi**
Untested for now.
This does not feature multi GPU, validation dataset or dataset consistency assessment.

### CODE REVIEW

**Comment by zaccharieramzi** [comment]
For now, I deactivated torch dense init. in this PR because of #54 .

**Comment by zaccharieramzi** [comment]
Atm I have 2 problems with TF ImageNet training:
- a memory leak (less severe when using `export MALLOC_TRIM_THRESHOLD_=0`, but still present
- very bad accuracy performances 
- very bad computational performances during evaluation : 40' for the training set.

Possible issues with performance are the classic : bad learning rate, wrong batch size.

As for the memory leak, it's very difficult to understand what's happening. I suspect something weird is going on either with the data augmentation, the evaluation process or the shuffling.

File: utils/torch_solver.py [hunk 1]
```diff
@@ -66,8 +66,12 @@ def set_objective(
...
 
         if self.data_aug:
+            if self.image_width < 128:
+                crop = transforms.RandomCrop(self.image_width, padding=4)
+            else:
+                crop = transforms.RandomResizedCrop(self.image_width)
             data_aug_list = [
-                transforms.RandomCrop(self.image_width, padding=4),
+                crop,
             ]
```
Comment (tomMoral) [inline utils/torch_solver.py:75]: Maybe easier to read if `data_aug_list = []` and you append the result of if in the list?

**Comment by zaccharieramzi** [comment]
Screenshot from 2022-11-18 18-26-47

For now we obtain 67.5% performance whereas we should be more around 71.5% with the pytorch example. There is definitely room for improvement but at least we are not completely off.
The differences I notice right away:
- different learning rate schedule (indeed it seems that we are decaying it too quickly towards the end)
- weight decay (5e-4 vs 1e-4 in pytorch)


The training time was 33 hours on a single V100 GPU.
I will relaunch one with these corrected parameters but for further investigation I would like to first have multi-gpu and amp working so that I can iterate faster.

Despite this number I think the PR is ready to be merged.

File: README.rst [hunk 2]
```diff
@@ -85,9 +86,10 @@ For example, to implement a new PyTorch-based solver with the Adam optimizer, yo
          optimizer = torch.optim.Adam(model.parameters(), lr=self.lr)
          # Initial evaluation
          while callback(model):
+            torch.cuda.empty_cache()
```
Comment (tomMoral) [inline README.rst:89]: If this is necessary, maybe add a comment?

File: datasets/imagenet.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    )
+
+
+class Dataset(MultiFrameworkDataset):
+
+    name = "Imagenet"
+
+    parameters = {
+        'framework': ['pytorch', 'lightning', 'tensorflow'],
+        'random_state': [42],
```
Comment (tomMoral) [inline datasets/imagenet.py:20]: Why is it necessary to have a seed? maybe put a comment?

File: datasets/imagenet.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        'MultiFrameworkDataset',
+    )
+
+
+class Dataset(MultiFrameworkDataset):
+
+    name = "Imagenet"
+
+    parameters = {
+        'framework': ['pytorch', 'lightning', 'tensorflow'],
```
Comment (tomMoral) [inline datasets/imagenet.py:19]: Do we need the `lightning` framework? or is it the same as the `pytorch` framework for datasets?

File: datasets/imagenet.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+    torch_split_kwarg = 'split'
+    torch_dl = False
+    extra_torch_test_transforms = [
+        transforms.Resize(256),
+        transforms.CenterCrop(224),
+    ]
+
+    tf_ds_name = 'imagenet2012'
+    tf_splits = ['validation', 'train']
```
Comment (tomMoral) [inline datasets/imagenet.py:52]: This order feels strange relative to the one l.55. If order is not important, can you reverse? if it is, can you add a comment?

File: objective.py [hunk 6]
```diff
@@ -104,7 +107,8 @@ def get_tf_model_init_fn(self):
         input_width = self.width
         if self.model_type == 'resnet':
             add_kwargs['use_bias'] = False
-            add_kwargs['dense_init'] = 'torch'
+            # for now deactivating
+            # add_kwargs['dense_init'] = 'torch'
```
Comment (tomMoral) [inline objective.py:111]: Should we deactivate this or not? if yes, can  you comment on why, why to keep it?

File: objective.py [hunk 7]
```diff
@@ -260,7 +295,10 @@ def compute(self, model):
         if self.with_validation:
             value_key = "val_err"
         else:
-            value_key = "train_loss"
+            if self.train_metrics:
+                value_key = "train_loss"
+            else:
+                value_key = "test_loss"
```
Comment (tomMoral) [inline objective.py:301]: I would say that for all cases, the value is `test_loss`, as this is what is used in most setups?

File: tests/test_dataset_consistency.py [hunk 8]
```diff
@@ -55,6 +57,10 @@ def __getitem__(self, idx):
         x, y = self.dataset[idx]
         if self.transform:
             x = self.transform(x)
+        try:
+            x = self.to_tensor(x)
+        except TypeError:
+            pass
```
Comment (tomMoral) [inline tests/test_dataset_consistency.py:63]: Can you comment on why this is necessary?

File: tests/test_tf_helper.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import tensorflow as tf
+
+from benchopt.utils.safe_import import set_benchmark
+
+set_benchmark('./')
```
Comment (tomMoral) [inline tests/test_tf_helper.py:5]: A safer way to set the benchmark is usually to use a relative location wrt `__file__`, so that it is independent of the runnign dir :
```suggestion
from pathlib import Path

set_benchmark(Path(__file__).parents[-2])
```

File: tests/test_tf_helper.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+
+def test_random_resized_crop():
+    from benchopt import safe_import_context
+    with safe_import_context() as import_ctx:
+        RandomResizedCrop = import_ctx.import_from(
+            'tf_helper', 'RandomResizedCrop'
+        )
```
Comment (tomMoral) [inline tests/test_tf_helper.py:13]: Wtih benchopt/benchopt#472, we can know use `from benchmark_utils.tf_helper import RandomResizedCrop`, without the context.

File: utils/tf_solver.py [hunk 11]
```diff
@@ -112,12 +115,21 @@ def set_objective(
...
+                )
+            else:
+                # random resize crop equivalent in tf
+                crop = RandomResizedCrop(
+                    scale=(0.08, 1.0),
+                    ratio=(0.75, 1.33),
+                    crop_shape=(self.image_width, self.image_width),
+                )
+            data_aug_list = [
+                tf.keras.layers.ZeroPadding2D(padding=4),
```
Comment (tomMoral) [inline utils/tf_solver.py:131]: Define this above the `if` and then append the correct crop. This avoids creating an extra local var just to append it in the sequence and keep a chronological definition.

**Review by tomMoral — APPROVED** [review]
A few comment but overall LGTM! Thx @zaccharieramzi :)
